exoscale/client.go

```diff
@@ -1,29 +1,162 @@
 	}
 
-	egoscale.UserAgent = fmt.Sprintf("Exoscale-K8s-Cloud-Controller/%s %s", versionString, egoscale.UserAgent)
+	return exoclient, nil
+}
+
+func (e *exoscaleClient) watchCredentialsFile(stop <-chan struct{}) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		fatalf("failed to watch credentials file %q, %q", e.credentialsFile, err)
+	}
+
+	// We watch the folder because the file might get deleted and recreated
+	err = watcher.Add(filepath.Dir(e.credentialsFile))
+	if err != nil {
+		fatalf("failed to watch credentials file %q, %q", e.credentialsFile, err)
+	}
+
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				errorf("credentials file watcher event channel closed")
+				return
+			}
+
+			if event.Name == e.credentialsFile &&
+				(event.Op&fsnotify.Write == fsnotify.Write || event.Op&fsnotify.Create == fsnotify.Create) {
+				infof("refreshing IAM credential from file %s", e.credentialsFile)
+				e.refreshCredentials()
+			}
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				errorf("credentials file watcher error channel closed")
+				return
+			}
+			errorf("error while watching credentials file %q, %q", e.credentialsFile, err)
+		case <-stop:
+			infof("closing credential file watch")
+			watcher.Close()
+			return
+		}
+	}
+
+}
+
+func (e *exoscaleClient) refreshCredentials() {
+	e.Lock()
+
+	f, err := os.Open(e.credentialsFile)
+	if err != nil {
+		fatalf("failed to read credentials file %q: %s", e.credentialsFile, err)
+	}
+
+	var credentials exoscaleAPICredentials
+	err = json.NewDecoder(f).Decode(&credentials)
+	if err != nil {
+		fatalf("failed to decode credentials file %q: %s", e.credentialsFile, err)
+	}
+	f.Close()
+
+	e.client = egoscale.NewClient(e.endpoint, credentials.APIKey, credentials.APISecret)
+	e.Unlock()
+
+	infof("exoscale IAM credentials refreshed, now using %s (%s)", credentials.Name, credentials.APIKey)
+}
+
+func (e *exoscaleClient) CreateNetworkLoadBalancer(ctx context.Context, zone string,
+	lbSpec *egoscale.NetworkLoadBalancer) (*egoscale.NetworkLoadBalancer, error) {
+	e.RLock()
+	defer e.RUnlock()
+
+	return e.client.CreateNetworkLoadBalancer(ctx, zone, lbSpec)
+}
+
+func (e *exoscaleClient) DeleteNetworkLoadBalancer(ctx context.Context, zone, id string) error {
+	e.RLock()
+	defer e.RUnlock()
+
+	return e.client.DeleteNetworkLoadBalancer(ctx, zone, id)
+}
+
+func (e *exoscaleClient) GetNetworkLoadBalancer(ctx context.Context, zone, id string) (*egoscale.NetworkLoadBalancer, error) {
+	e.RLock()
+	defer e.RUnlock()
+
+	return e.client.GetNetworkLoadBalancer(ctx, zone, id)
+}
+
+func (e *exoscaleClient) UpdateNetworkLoadBalancer(ctx context.Context, zone string,
+	nlbUpdate *egoscale.NetworkLoadBalancer) (*egoscale.NetworkLoadBalancer, error) {
+	e.RLock()
+	defer e.RUnlock()
+
+	return e.client.UpdateNetworkLoadBalancer(ctx, zone, nlbUpdate)
+}
+
+func (e *exoscaleClient) GetVirtualMachine(ctx context.Context, uuid *egoscale.UUID) (*egoscale.VirtualMachine, error) {
```

Comment:
I know that the underlying type is still `egoscale.VirtualMachine` for historical reasons (which will change in API v2), however we do our best to normalize the name everywhere as *[Compute] instance*:

```suggestion
func (e *exoscaleClient) GetInstance(ctx context.Context, uuid *egoscale.UUID) (*egoscale.VirtualMachine, error) {
```